Merge `RequestHandler` and `ErrorHandler` interfaces

Technically everywhere in Express can accept both and it just skips over it when the arity is wrong. For example, `app.get(handler, errorHandler)` is valid.

You are probably right, it doesn't seem like it from the documentation though. I tried it like this:

``` ts
    export type Handler = RequestHandler | ErrorHandler;

    /** First argument to all routing methods like `use`, `get`, `all` etc */
    export type PathParams = string | RegExp | (string | RegExp)[];

    /** Can be passed to all methods like `use`, `get`, `all` etc */
    export type HandlerArgument = Handler | Handler[];
```

Which is definitely cleaner and makes for shorter method declarations, but when using this, the types for `req`, `res` and `next` are not inferred anymore:

```
test/router.ts(6,27): error TS7006: Parameter 'req' implicitly has an 'any' type.
test/router.ts(6,32): error TS7006: Parameter 'res' implicitly has an 'any' type.
test/router.ts(11,11): error TS7006: Parameter 'req' implicitly has an 'any' type.
test/router.ts(11,16): error TS7006: Parameter 'res' implicitly has an 'any' type.
test/router.ts(11,21): error TS7006: Parameter 'next' implicitly has an 'any' type.
test/app.ts(15,26): error TS7006: Parameter 'req' implicitly has an 'any' type.
test/app.ts(15,31): error TS7006: Parameter 'res' implicitly has an 'any' type.
test/app.ts(15,36): error TS7006: Parameter 'next' implicitly has an 'any' type.
test/app.ts(22,28): error TS7006: Parameter 'req' implicitly has an 'any' type.
test/app.ts(22,33): error TS7006: Parameter 'res' implicitly has an 'any' type.
test/app.ts(22,38): error TS7006: Parameter 'next' implicitly has an 'any' type.
```

What do you think? Is it possible to type and keep inferring and still use the type aliases? Or should we just require the user to explicitly type?

--- a/lib/application.d.ts
+++ b/lib/application.d.ts
@@ -2,7 +2,7 @@
 import {Server} from 'http';
 import {ListenOptions} from 'net';
 import {EventEmitter} from 'events';
-import {Router, RequestHandler, ParamHandler, PathArgument} from './router/index';
+import {Router, RequestHandler, ParamHandler, HandlerArgument, PathArgument} from './router/index';
 
 declare namespace app {
     export interface Application extends EventEmitter, Router {
@@ -81,7 +81,7 @@ declare namespace app {
         set(setting: string, val: any): this;
         get(name: string): any;
         // need to duplicate this here from the Router because of the overload
-        get(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
+        get(path: PathArgument, ...handlers: HandlerArgument[]): this;
 
         /**
          * Add callback triggers to route parameters, where name is the name of the parameter or an array of them,

--- a/lib/express.d.ts
+++ b/lib/express.d.ts
@@ -16,6 +16,7 @@ declare namespace createApplication {
     export interface RequestHandler extends createRouter.RequestHandler { }
     export interface ErrorHandler extends createRouter.ErrorHandler { }
     export interface ParamHandler extends createRouter.ParamHandler { }
+    export type Handler = createRouter.Handler
     export interface NextFunction extends createRouter.NextFunction { }
 
     // need to use an interface for this because `static` is a reserved word

--- a/lib/router/index.d.ts
+++ b/lib/router/index.d.ts
@@ -19,6 +19,11 @@ declare namespace createRouter {
         (err: any, req: Request, res: Response, next: NextFunction): any;
     }
 
+    export type Handler = RequestHandler | ErrorHandler;
+
+    /** Can be passed to all methods like `use`, `get`, `all` etc */
+    export type HandlerArgument = Handler | Handler[];
+
     export interface ParamHandler {
         (req: Request, res: Response, next: NextFunction, value: any, name: string): any;
     }
@@ -58,36 +63,36 @@ declare namespace createRouter {
          */
         param(callback: (name: string, matcher: RegExp) => ParamHandler): this;
 
-        all(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        get(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        post(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        put(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        head(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        delete(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        options(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        trace(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        copy(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        lock(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        mkcol(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        move(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        purge(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        propfind(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        proppatch(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        unlock(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        report(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        mkactivity(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        checkout(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        merge(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        'm-search'(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        notify(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        subscribe(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        unsubscribe(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        patch(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        search(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
-        connect(path: PathArgument, ...handlers: (RequestHandler | RequestHandler[])[]): this;
+        all(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        get(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        post(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        put(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        head(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        delete(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        options(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        trace(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        copy(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        lock(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        mkcol(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        move(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        purge(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        propfind(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        proppatch(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        unlock(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        report(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        mkactivity(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        checkout(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        merge(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        'm-search'(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        notify(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        subscribe(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        unsubscribe(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        patch(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        search(path: PathArgument, ...handlers: HandlerArgument[]): this;
+        connect(path: PathArgument, ...handlers: HandlerArgument[]): this;
 
-        use(...handlers: (RequestHandler | ErrorHandler | (RequestHandler | ErrorHandler)[])[]): this;
-        use(mountPoint: string, ...handlers: (RequestHandler | ErrorHandler | (RequestHandler | ErrorHandler)[])[]): this;
+        use(...handlers: HandlerArgument[]): this;
+        use(mountPoint: string, ...handlers: HandlerArgument[]): this;
 
         route(prefix: PathArgument): Route;
     }

--- a/lib/router/route.d.ts
+++ b/lib/router/route.d.ts
@@ -1,5 +1,5 @@
 
-import {RequestHandler} from './index';
+import {HandlerArgument} from './index';
 
 declare class Route {
 
@@ -8,33 +8,33 @@ declare class Route {
 
     constructor(path: string);
 
-    all(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    get(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    post(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    put(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    head(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    delete(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    options(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    trace(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    copy(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    lock(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    mkcol(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    move(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    purge(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    propfind(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    proppatch(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    unlock(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    report(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    mkactivity(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    checkout(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    merge(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    'm-search'(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    notify(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    subscribe(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    unsubscribe(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    patch(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    search(...handlers: (RequestHandler | RequestHandler[])[]): this;
-    connect(...handlers: (RequestHandler | RequestHandler[])[]): this;
+    all(...handlers: HandlerArgument[]): this;
+    get(...handlers: HandlerArgument[]): this;
+    post(...handlers: HandlerArgument[]): this;
+    put(...handlers: HandlerArgument[]): this;
+    head(...handlers: HandlerArgument[]): this;
+    delete(...handlers: HandlerArgument[]): this;
+    options(...handlers: HandlerArgument[]): this;
+    trace(...handlers: HandlerArgument[]): this;
+    copy(...handlers: HandlerArgument[]): this;
+    lock(...handlers: HandlerArgument[]): this;
+    mkcol(...handlers: HandlerArgument[]): this;
+    move(...handlers: HandlerArgument[]): this;
+    purge(...handlers: HandlerArgument[]): this;
+    propfind(...handlers: HandlerArgument[]): this;
+    proppatch(...handlers: HandlerArgument[]): this;
+    unlock(...handlers: HandlerArgument[]): this;
+    report(...handlers: HandlerArgument[]): this;
+    mkactivity(...handlers: HandlerArgument[]): this;
+    checkout(...handlers: HandlerArgument[]): this;
+    merge(...handlers: HandlerArgument[]): this;
+    'm-search'(...handlers: HandlerArgument[]): this;
+    notify(...handlers: HandlerArgument[]): this;
+    subscribe(...handlers: HandlerArgument[]): this;
+    unsubscribe(...handlers: HandlerArgument[]): this;
+    patch(...handlers: HandlerArgument[]): this;
+    search(...handlers: HandlerArgument[]): this;
+    connect(...handlers: HandlerArgument[]): this;
 }
 
 export = Route;

--- a/test/app.ts
+++ b/test/app.ts
@@ -12,14 +12,13 @@ const requestHandler: express.RequestHandler = (req, res, next) => {
     next(new Error());
 };
 
-app.get('/hello-world', (req, res, next) => {
-    // inferred types
+app.get('/hello-world', (req: express.Request, res: express.Response, next: express.NextFunction) => {
     console.log(req.hostname);
     res.send('Hello World');
     next();
 });
 
-app.put('/users/:userId', (req, res, next) => {
+app.put('/users/:userId', (req: express.Request, res: express.Response, next: express.NextFunction) => {
     res.status(400);
 });
 

--- a/test/router.ts
+++ b/test/router.ts
@@ -1,14 +1,14 @@
 
-import {Router} from 'express';
+import {Router, Request, Response, NextFunction} from 'express';
 
 const router = Router();
 
-router.move('/somepath', (req, res) => {
+router.move('/somepath', (req: Request, res: Response) => {
     res.set('Location', res.get('Destination'));
 });
 
 router.route('/someroute')
-    .get((req, res, next) => {
+    .get((req: Request, res: Response, next: NextFunction) => {
         req.accepts('application/json');
         console.log('hello');
     });

